In [1]:
import importlib
import plotly.figure_factory as ff
import src.plot_data
importlib.reload(src.plot_data)

from src.plot_data import plot_data

for players in [*range(2, 11)]:
    plot_data(show_plot=True, n_players_to_plot=players)

INFO: Making plotly figure and axis objects
c:\Users\david\Documents\Projects\holdem_simulator\src\plot_builders_plotly.py:158: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataframe[y3_var] = dataframe[y3_var].round(4)
INFO: Sorting data frame ascending by n_players
INFO: Sorting data frame descending by win ratio, win ratio rounded down to nearest 5%
INFO: Showing plot object


INFO: Not saving plot - pass save_plot=True to save it.
INFO: Making plotly figure and axis objects
c:\Users\david\Documents\Projects\holdem_simulator\src\plot_builders_plotly.py:158: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

INFO: Sorting data frame ascending by n_players
INFO: Sorting data frame descending by win ratio, win ratio rounded down to nearest 5%
INFO: Showing plot object


INFO: Not saving plot - pass save_plot=True to save it.
INFO: Making plotly figure and axis objects
c:\Users\david\Documents\Projects\holdem_simulator\src\plot_builders_plotly.py:158: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

INFO: Sorting data frame ascending by n_players
INFO: Sorting data frame descending by win ratio, win ratio rounded down to nearest 5%
INFO: Showing plot object


INFO: Not saving plot - pass save_plot=True to save it.
INFO: Making plotly figure and axis objects
c:\Users\david\Documents\Projects\holdem_simulator\src\plot_builders_plotly.py:158: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

INFO: Sorting data frame ascending by n_players
INFO: Sorting data frame descending by win ratio, win ratio rounded down to nearest 5%
INFO: Showing plot object


INFO: Not saving plot - pass save_plot=True to save it.
INFO: Making plotly figure and axis objects
c:\Users\david\Documents\Projects\holdem_simulator\src\plot_builders_plotly.py:158: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

INFO: Sorting data frame ascending by n_players
INFO: Sorting data frame descending by win ratio, win ratio rounded down to nearest 5%
INFO: Showing plot object


INFO: Not saving plot - pass save_plot=True to save it.
INFO: Making plotly figure and axis objects
c:\Users\david\Documents\Projects\holdem_simulator\src\plot_builders_plotly.py:158: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

INFO: Sorting data frame ascending by n_players
INFO: Sorting data frame descending by win ratio, win ratio rounded down to nearest 5%
INFO: Showing plot object


INFO: Not saving plot - pass save_plot=True to save it.
INFO: Making plotly figure and axis objects
c:\Users\david\Documents\Projects\holdem_simulator\src\plot_builders_plotly.py:158: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

INFO: Sorting data frame ascending by n_players
INFO: Sorting data frame descending by win ratio, win ratio rounded down to nearest 5%
INFO: Showing plot object


INFO: Not saving plot - pass save_plot=True to save it.
INFO: Making plotly figure and axis objects
c:\Users\david\Documents\Projects\holdem_simulator\src\plot_builders_plotly.py:158: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

INFO: Sorting data frame ascending by n_players
INFO: Sorting data frame descending by win ratio, win ratio rounded down to nearest 5%
INFO: Showing plot object


INFO: Not saving plot - pass save_plot=True to save it.
INFO: Making plotly figure and axis objects
c:\Users\david\Documents\Projects\holdem_simulator\src\plot_builders_plotly.py:158: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

INFO: Sorting data frame ascending by n_players
INFO: Sorting data frame descending by win ratio, win ratio rounded down to nearest 5%
INFO: Showing plot object


INFO: Not saving plot - pass save_plot=True to save it.


In [2]:
plot_data(show_plot=True, n_players_to_plot="All")

INFO: Making plotly figure and axis objects
c:\Users\david\Documents\Projects\holdem_simulator\src\plot_builders_plotly.py:158: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

INFO: Sorting data frame ascending by n_players
INFO: Sorting data frame descending by win ratio, win ratio rounded down to nearest 5%


INFO: Showing plot object


INFO: Not saving plot - pass save_plot=True to save it.


In [3]:
from src.plot_data import _load_all_players_wins_by_hole_cards_flavor_df

df = _load_all_players_wins_by_hole_cards_flavor_df()
df = df.drop(
    columns=["appearances", "wins", "win ratio", "fewer than expected appearances"]
)
df.head()

hole cards flavor  win ratio rounded down to nearest 5%  n_players
0       Aces paired                                  0.85          2
1      Kings paired                                  0.80          2
2     Queens paired                                  0.75          2
3      Jacks paired                                  0.75          2
4        10s paired                                  0.75          2

In [36]:
df.shape

(1521, 3)

In [32]:
def make_heatmap_for_hole_card_families(str_to_search: str):
    tmp_df = (
        df[df["hole cards flavor"].str.contains(str_to_search)]
        .pivot_table(
            index="hole cards flavor",
            columns="n_players",
            values="win ratio rounded down to nearest 5%",
        )
        .rename(
            columns={
                2: "2",
                3: "3",
                4: "4",
                5: "5",
                6: "6",
                7: "7",
                8: "8",
                9: "9",
                10: "10",
            }
        )
        .reset_index()
        .rename_axis(None, axis=1)
    )

    for i in range(2, 10):
        tmp_df[f"Prob to win with {i} players > {i+1} players"] = (
            tmp_df[str(i)] - tmp_df[str(i + 1)]
        )

    tmp_df = tmp_df.sort_values(
        by=[str(i) for i in range(2, 11)], ascending=False
    ).drop(columns=[str(i) for i in range(2, 11)])

    tmp_df.set_index(df.columns[0], inplace=True)

    tmp_df = tmp_df.round(2)

    y_labels = tmp_df.index.tolist()[::-1]

    x_labels = tmp_df.columns.tolist()

    z_values = tmp_df.values[::-1]

    fig = ff.create_annotated_heatmap(z=z_values, x=x_labels, y=y_labels)

    fig.show()

    # Add print statement that shows the median value for each column
    print(tmp_df.median())

In [33]:
make_heatmap_for_hole_card_families(str_to_search="paired")

Prob to win with 2 players > 3 players     0.20
Prob to win with 3 players > 4 players     0.10
Prob to win with 4 players > 5 players     0.05
Prob to win with 5 players > 6 players     0.05
Prob to win with 6 players > 7 players     0.05
Prob to win with 7 players > 8 players     0.05
Prob to win with 8 players > 9 players     0.00
Prob to win with 9 players > 10 players    0.00
dtype: float64


## For paired hole cards

- For 8-10 players, use a raw score.
- If 7 players, add 5%
- If 6 players, add another 5%
- If 5 players, add another 5%
- If 4 players, add another 5%
- If 3 players, add another 10%
- If 2 players, add another 20%

In [34]:
make_heatmap_for_hole_card_families(str_to_search="suited")

Prob to win with 2 players > 3 players     0.15
Prob to win with 3 players > 4 players     0.10
Prob to win with 4 players > 5 players     0.05
Prob to win with 5 players > 6 players     0.05
Prob to win with 6 players > 7 players     0.00
Prob to win with 7 players > 8 players     0.00
Prob to win with 8 players > 9 players     0.00
Prob to win with 9 players > 10 players    0.00
dtype: float64


## For suited flavors

- For 6-10 players, use a raw score.
- If 5 players, add 5%
- If 4 players, add another 5%
- If 3 players, add another 10%
- If 2 players, add another 15%

In [35]:
make_heatmap_for_hole_card_families(str_to_search="off suit")

Prob to win with 2 players > 3 players     0.15
Prob to win with 3 players > 4 players     0.10
Prob to win with 4 players > 5 players     0.05
Prob to win with 5 players > 6 players     0.05
Prob to win with 6 players > 7 players     0.05
Prob to win with 7 players > 8 players     0.00
Prob to win with 8 players > 9 players     0.00
Prob to win with 9 players > 10 players    0.00
dtype: float64


## For off suit flavors

- For 7-10 players, use a raw score.
- If 6 players, add 5%
- If 5 players, add another 5%
- If 4 players, add another 5%
- If 3 players, add another 10%
- If 2 players, add another 15%

# The heuristic to build

1. Come up with scores to use for 10-player games based on ranks of hole cards, if they pair, if they have flush potential, if they have straight potential that yield the following:
  - Pocket Aces: 30%
  - Pocket 2s: 10%
  - Ace, King suited: 20%
  - 3, 2 suited: 5%
  - Ace, King off suit: 15%
  - 3, 2 off suit: 5%
2. For every player fewer than 8 players, add 5% to the score. This will result in a 30% bonus applied for heads-up games so that the numbers would become:
  - Pocket Aces: 60%
  - Pocket 2s: 40%
  - Ace, King suited: 50%
  - 3, 2 suited: 35%
  - Ace, King off suit: 45%
  - 3, 2 off suit: 35%
3. If there are fewer than 8 players, add another 10% to the score if holding a pocket pair so that the numbers would become (in a heads-up game):
  - Pocket Aces: 70%
  - Pocket 2s: 50%
  - Ace, King suited: 50%
  - Ace, King off suit: 55%
4. If there are fewer than 8 players, add 5% for every card that is rank 10 or greater so that the numbers would become (in a heads-up game):
  - Pocket Aces: 80%
  - Ace, King suited: 60%
  - Ace, King off suit: 65%
5. For now, live with the fact that your heuristic is conservative and does not give you the final 5% for some of these edge cases in heads-up games:
  - Pocket Aces: 85% (5% better than currently arrived at)
  - Ace, King suited: 65% (5% better than currently arrived at)

# The training program to build

Now that you have the probabilities of winning for every hole card flavor for every player count from 2-10 rounded down to the nearest 5%, you can look those up every time you do a practice hand. As long as you do not over-estimate the strength of your hand (and therefore over commit), you will be getting good training. The df created by `_load_all_players_wins_by_hole_cards_flavor_df` is a good starting point for creating this file.